src/main/java/org/springframework/integration/aws/outbound/KinesisMessageHandler.java

```diff
@@ -245,4 +297,44 @@ private PutRecordRequest buildPutRecordRequest(Message<?> message) {
 				.withData(data);
 	}
 
+	@SuppressWarnings("rawtypes")
+	private AsyncHandler<? extends AmazonWebServiceRequest, ?> getAsyncHandler(final Message<?> message,
+																			final AmazonWebServiceRequest request) {
+		if (this.asyncHandler != null) {
+			return this.asyncHandler;
+		}
+		else {
+			return new AsyncHandler<AmazonWebServiceRequest, AmazonWebServiceResult>() {
+
+				@Override
+				public void onError(Exception ex) {
+					if (getSendFailureChannel() != null) {
+						KinesisMessageHandler.this.messagingTemplate.send(getSendFailureChannel(),
+								KinesisMessageHandler.this.errorMessageStrategy.buildErrorMessage(
+										new AwsRequestFailureException(message, request, ex), null));
+					}
+				}
+
+				@Override
+				public void onSuccess(AmazonWebServiceRequest request, AmazonWebServiceResult result) {
+					Message<?> resultMessage;
+
+					if (result instanceof PutRecordResult) {
+						resultMessage = getMessageBuilderFactory().fromMessage(message)
+								.setHeader(AwsHeaders.SHARD, ((PutRecordResult) result).getShardId())
+								.setHeader(AwsHeaders.SEQUENCE_NUMBER, ((PutRecordResult) result).getSequenceNumber())
+								.build();
+					}
+					else {
+						resultMessage = getMessageBuilderFactory().fromMessage(message).build();
```

Comment:
In this else I guess we should deal with the `PutRecordsResult` and populate appropriate header on the matter. Maybe a new one like `serviceResult` into the `AwsHeaders` and set the full result object...